lib/api/core/httpRouter/index.js

```diff
@@ -158,48 +158,58 @@ class Router {
           .catch(error => replyWithError(cb, request, error));
       }
 
-      return replyWithError(cb, request, new BadRequestError(`Unrecognized HTTP method ${httpRequest.method}`));
+      return Promise.reject(replyWithError(cb, request, new BadRequestError(`Unrecognized HTTP method ${httpRequest.method}`)));
     }
 
     httpRequest.url = httpRequest.url.replace(LeadingSlashRegex, '');
 
     try {
       const routeHandler = this.routes[httpRequest.method].getHandler(httpRequest);
+      const httpEvent = `http:${httpRequest.method.toLowerCase()}`;
 
       request = routeHandler.getRequest();
       request.response.setHeaders(this.defaultHeaders);
 
-      if (routeHandler.handler === null) {
-        return replyWithError(cb, request, new NotFoundError(`API URL not found: ${routeHandler.url}`));
-      }
-
-      if (httpRequest.content.length <= 0) {
-        return routeHandler.invokeHandler(cb);
-      }
-
-      if (httpRequest.headers['content-type']
-        && !httpRequest.headers['content-type'].startsWith('application/json')) {
-        return replyWithError(cb, request, new BadRequestError(`Invalid request content-type. Expected "application/json", got: "${httpRequest.headers['content-type']}"`));
-      }
+      return this.kuzzle.pluginsManager.trigger(httpEvent, request)
```

Comment:
These new events must be triggered **after** the request has been validated. It makes no sense sending a request to plugins if an URL is invalid or if a request is malformed.